### PR TITLE
[spirv] Fully translate pass-through pixel shader!

### PR DESCRIPTION
**Pull request by antiagainst**
* Add support for generating stage IO variables
* Add support for interface variable ids in entry point
* Add support for OpReturnValue

### CODE REVIEW

**Comment by antiagainst** [comment]
This PR still needs cleanup (adding comments, renaming variables/methods, etc.).

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h [hunk 1]
```diff
@@ -32,8 +32,9 @@ class ModuleBuilder {
   /// \brief Begins building a SPIR-V function. At any time, there can only
   /// exist at most one function under building. Returns the <result-id> for the
   /// function on success. Returns zero on failure.
-  uint32_t beginFunction(uint32_t funcType, uint32_t returnType,
-                         const std::vector<uint32_t> &paramTypeIds = {});
+  uint32_t beginFunction(uint32_t funcType, uint32_t returnType);
+
+  uint32_t addFnParameter(uint32_t type);
```
Comment (ehsannas) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: add doc: `Registers the function parameter and returns the <result-id> for it`

Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: Done

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h [hunk 2]
```diff
@@ -42,16 +43,21 @@ class ModuleBuilder {
...
-  bool bbReturn(uint32_t labelId);
+  /// \brief Returns true if the current basic block containing the insertion
+  /// point is terminated.
+  inline bool isCurBasicBlockTerminated() const;
 
   /// \brief Sets insertion point to the basic block with the given <label-id>.
   /// Returns true on success, false on failure.
   bool setInsertPoint(uint32_t labelId);
 
+  uint32_t createLoad(uint32_t resultType, uint32_t pointer);
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: Done

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h [hunk 3]
```diff
@@ -73,6 +79,11 @@ class ModuleBuilder {
   uint32_t getVec4Type(uint32_t elemType);
   uint32_t getFunctionType(uint32_t returnType,
                            const std::vector<uint32_t> &paramTypes);
+  uint32_t getPointerType(uint32_t pointeeType, spv::StorageClass);
+
+  uint32_t addStageIOVariable(uint32_t type, spv::StorageClass storageClass,
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: Done

File: tools/clang/include/clang/SPIRV/ModuleBuilder.h [hunk 4]
```diff
@@ -93,6 +104,10 @@ class ModuleBuilder {
   InstBuilder instBuilder;
 };
 
+bool ModuleBuilder::isCurBasicBlockTerminated() const {
+  return insertPoint->isTerminated();
```
Comment (antiagainst) [inline tools/clang/include/clang/SPIRV/ModuleBuilder.h]: Yep, good catch! Done

File: tools/clang/lib/SPIRV/ModuleBuilder.cpp [hunk 5]
```diff
@@ -37,15 +35,19 @@ ModuleBuilder::beginFunction(uint32_t funcType, uint32_t returnType,
...
-  // instruction per each formal parameter of the function.
-  for (const auto &typeId : paramTypeIds) {
-    theFunction->addParameter(typeId, theContext.takeNextId());
-  }
-
   return fId;
 }
 
+uint32_t ModuleBuilder::addFnParameter(uint32_t type) {
+  if (!theFunction) {
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/ModuleBuilder.cpp]: Done.

File: tools/clang/lib/SPIRV/ModuleBuilder.cpp [hunk 6]
```diff
@@ -80,26 +82,40 @@ uint32_t ModuleBuilder::bbCreate() {
   return labelId;
 }
 
-bool ModuleBuilder::bbReturn(uint32_t labelId) {
+bool ModuleBuilder::setInsertPoint(uint32_t labelId) {
   auto it = basicBlocks.find(labelId);
   if (it == basicBlocks.end()) {
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/ModuleBuilder.cpp]: Hmm, it depends how we define the API in `ModuleBuilder`. There are two approaches currently in `ModuleBuilder`:

* Return on valid execution path and explode on invalid usage (violating assumptions). Various instruction insertion methods are following this.
* Return valid ids on valid execution path and return 0 on invalid usage. Function/BasicBlock creating/ending methods are following this.

I kinda think `setInsertPoint` belongs to the second case.

But it also makes sense to unify these two approaches.

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 7]
```diff
@@ -22,22 +23,35 @@
...
-  // Profile (DXIL Model) : HLSL Shader Kind : SPIR-V Shader Kind
+  // Profile (DXIL Model) : HLSL Shader Kind : SPIR-V Shader Stage
   // vs_<version>         : Vertex Shader    : Vertex Shader
   // hs_<version>         : Hull Shader      : Tassellation Control Shader
   // ds_<version>         : Domain Shader    : Tessellation Evaluation Shader
   // gs_<version>         : Geometry Shader  : Geometry Shader
   // ps_<version>         : Pixel Shader     : Fragment Shader
   // cs_<version>         : Compute Shader   : Compute Shader
+
+  if (!profile) {
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Done

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 8]
```diff
@@ -49,11 +63,268 @@ spv::ExecutionModel getSpirvShaderStageFromHlslProfile(const char *profile) {
...
+/// corresponding SPIR-V <result-id>s.
+///
+/// All symbols defined in the AST should be "defined" or registered in this
+/// class and have their <result-id>s queried from this class. In the process
+/// of defining a Decl, the SPIR-V module builder passed into the constructor
+/// will be used to generate all SPIR-V instructions required.
+///
+/// This class acts as a middle layer to handle the mapping between HLSL
+/// semantics and Vulkan interface (stage builtin/input/output) variables.
+/// Such mapping is required because the semantic differences between DirectX
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Yep! Done

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 9]
```diff
@@ -49,11 +63,268 @@ spv::ExecutionModel getSpirvShaderStageFromHlslProfile(const char *profile) {
...
+  uint32_t defineFnParam(ParmVarDecl *paramDecl) {
+    const auto sk = getInterfaceVarSemanticAndKind(paramDecl);
+    if (sk.second != InterfaceVariableKind::None) {
+      // Found parameter with semantic attached. This means we need to map the
+      // parameter to a single interface variable.
+      const uint32_t paramTypeId =
+          translateType(paramDecl->getType(), theBuilder);
+      // TODO: Change to the correct interface variable kind here.
+      const uint32_t varId = theBuilder.addStageIOVariable(
+          paramTypeId, spv::StorageClass::Input, llvm::None);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Yep, it truly depends. Not all parameters are inputs to the shader. I've TODO in the above.

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 10]
```diff
@@ -49,11 +63,268 @@ spv::ExecutionModel getSpirvShaderStageFromHlslProfile(const char *profile) {
...
+
+      stageInputs.push_back(std::make_pair(varId, sk.first));
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Yes, definitely need to revise this later.

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 11]
```diff
@@ -49,11 +63,268 @@ spv::ExecutionModel getSpirvShaderStageFromHlslProfile(const char *profile) {
...
+    None,   ///< Not an interface variable
+    Input,  ///< Stage input variable
+    Output, ///< Stage output variable
+    IO, ///< Interface variable that can act as both stage input or stage output
+    // TODO: builtins
+  };
+
+  using InterfaceVarIdSemanticPair = std::pair<uint32_t, llvm::StringRef>;
+
+  /// \brief Returns the interface variable kind for the given Decl.
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Done

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 12]
```diff
@@ -112,76 +393,120 @@ class SPIRVEmitter : public ASTConsumer {
...
-
-  uint32_t translateFunctionType(FunctionDecl *decl,
-                                 std::vector<uint32_t> *paramTypes) {
-    const uint32_t retType = translateType(decl->getReturnType());
+    curFunction = decl;
+
+    uint32_t retType = declIdMapper.defineFnReturn(decl);
+
+    // Process function parameters. We need to strip parameters mapping to stage
+    // builtin/input/output variables and use what left in the function type.
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Done

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp [hunk 13]
```diff
@@ -112,76 +393,120 @@ class SPIRVEmitter : public ASTConsumer {
...
+  void doStmt(Stmt *stmt) {
+    if (auto *retStmt = dyn_cast<ReturnStmt>(stmt)) {
+      doReturnStmt(retStmt);
+    }
+    // TODO: handle other statements
+  }
+
+  void doReturnStmt(ReturnStmt *stmt) {
+    const uint32_t retValue = doExpr(stmt->getRetValue());
+    const uint32_t ifxVarId = declIdMapper.getRemappedDeclResultId(curFunction);
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Changed to `interface`

File: tools/clang/lib/SPIRV/Structure.cpp [hunk 14]
```diff
@@ -50,15 +50,19 @@ void BasicBlock::take(InstBuilder *builder) {
...
   builder->opLabel(labelId).x();
   for (auto &inst : instructions) {
     builder->getConsumer()(std::move(inst));
   }
   clear();
 }
 
+bool BasicBlock::isTerminated() const {
+  return !instructions.empty() && isTerminator(static_cast<spv::Op>(
+                                      instructions.back().front() & 0xffff));
```
Comment (antiagainst) [inline tools/clang/lib/SPIRV/Structure.cpp]: Done

File: tools/clang/lib/SPIRV/EmitSPIRVAction.cpp (hunk 9 above)
Comment (ehsannas) [inline tools/clang/lib/SPIRV/EmitSPIRVAction.cpp]: Oh I missed the TODO there. thanks.

**Pull request merged**